### PR TITLE
Explore: Display all values in graph's tooltip

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What is this feature?**

Currently Explore's graph is showing only 1 value in tooltip (the hovered-over one). However, showing all values is probably more useful, as users can see all values without needing to hover over multiple lines. Moreover, log volume is already using multi tooltip mode.

After (with multimode):

image

Before (single mode):

image

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/explore/Graph/ExploreGraph.tsx
Comment (Elfo404) [inline public/app/features/explore/Graph/ExploreGraph.tsx:72]: can we just remove this prop and cleanup if we are always going to use multi?

Comment (ivanahuckova) [reply public/app/features/explore/Graph/ExploreGraph.tsx:72]: Yeah, let's do that. New commit coming nooooow....

**Comment by ifrost** [comment]
Mmmm... I'm not sure about this one. Unlike in dashboards, the user has no control over the tooltip mode. We use `Single` mode because it looks good with just a few lines but looks like a mess with more lines:

Screenshot 2023-01-12 at 21 28 10

For a person with color recognition problems that makes the tooltip totally useless :D

I think we need to call Zordon. @radiorental?

**Comment by ivanahuckova** [comment]
> For a person with color recognition problems that makes the tooltip totally useless :D

But isn't the whole graph a problem then? I'd argue that in that case, you would click on legend of metric you want to see and hover over only that tooltip. 

image

**Comment by ifrost** [comment]
> But isn't the whole graph a problem then? I'd argue that in that case, you would click on legend of metric you want to see and hover over only that tooltip. Moreover you have a name next to it.

For me, it's impossible to use the legend to match the color of the line, though I just noticed that the line you hover over is emphasized with bold text inside the tooltip! It's hard to notice but should be good enough.

**Comment by leeoniya** [comment]
slightly off-topic, but i'm not sure why the TimeSeries panel defaults to single mode. i think Graph (old) default was All and seems like a better default.

**Review by Elfo404 — APPROVED** [review]
🔥

**Review by ifrost — APPROVED** [review]
It looks good 👍 I still would like @radiorental to say if it's a good change UX-wise before merging it. I'm not fully convinced that `Multi` trumps `Single` in all cases when it cannot be controlled by the user. To me, it actually adds more noise, but maybe I just got used to it. BTW. the default option in Dashboards seems to be `Single`.

**Comment by radiorental** [comment]
i think the mutli solution as I understand it does not scale.  I *think* users would want to compare 2 results / datapoints in general so over a set of results / traces the solution here is to sticky the primary trace of interest and then hover over other traces with the horizontal axis fixed.  

This requires a little bit of design to afford to the user they can sticky the x-axis. 

Bottom line, I would advise against this both in terms of scalability and accessibility 
image

**Comment by ifrost** [comment]
> They might want to compare values of 2-X series over time

I think that's probably the most frequent use case. Maybe we could switch to `multi` depending on the number of shown series and after some threshold fall back to single mode as you mentioned in:

> And also, if readability/noise of tooltip in graph with 10+ lines is the main problem, couldn't we just use single for graph with 10+ series?

Not sure if 10 is the best threshold 🤷  WDYT @radiorental?

>Imagine that you would be supposed to find out top 5 series at 20:32 and 21:17. I'd say tooltip is the only way you would be able to achieve this.

Not sure how would you do it - the tooltip doesn't seem to be sorted by value but by name 🤔  Also not sure why would you use a tooltip to see top/bottom X series, I think most query languages support sorting/filtering.

> First of all, is graph with 10+ lines even usable/understandable and common usecase? And should it be our main concern and focus?

Yes, a very common use case is when you see metrics for a lot of series but only one stands out (e.g. error rate)

**Comment by ivanahuckova** [comment]
>Yes, a very common use case is when you see metrics for a lot of series but only one stands out (e.g. error rate)

Yeah, thats a great point. I can see how multi tooltip could be in a way for this usecase.

**Comment by ifrost** [comment]
I think having a threshold to switch between `single` / `multi` could be a good compromise. Another food for thought: it could be nice to have a keyboard shortcut to quickly switch between both.

**Comment by radiorental** [comment]
Can someone give me some real world use cases for comparing more than 2 series vs 2.  Which is more likely?  If there's an 80/20 rule here where the majority of the time its compare A against B and sometimes it's A against n then there are a few UI patterns to solve for that more elegantly than just displaying all or 10.

I'm not a fan of an arbitrary threshold because that then needs to be communicated to users who otherwise may perceive that as a bug, i.e. there's 13 traces and the tooltip shows values  for 10, why?

We should aim to avoid pushing control of what is displayed onto the user, it incurs UX debt in the sense that it's more UI somewhere that adds to the overall clutter we already suffer from.

**Comment by ivanahuckova** [comment]
> there's 13 traces and the tooltip shows values for 10, why?

I think the idea was to show single tooltip with 1 value if there is more than N number of series. And show multi tooltip with all values if there is less than N number of series. N being specified number - 2 or 5 or 10 or more...

**Comment by radiorental** [comment]
Thanks for the clarification but I think the underlying UX issue still stands:  Inconsistent behaviour to the uninformed eye.  Lets say we set the the cut off to 10.  In the case where there are two sets of results (either on the screen at the same time or separately) The first has 5 traces, the second has 15.  Imho I would be concerned that the view would perceive a bug in the second visualisaton.

With that all said, I feel we're in the realm of speculation.  Can we ground the conversation with real world use cases for how people are utilising the information in the tooltip.  Does anyone need to see 10 values in parallel?  My best guess is that people only ever compare 2 values.  If that guess is valid the solution here is to figure out how to enable that A:B comparison, not just try to dump as many values onto the screen as elegantly as possible and leave it to the user to find the data they're interested in.

**Comment by radiorental** [comment]
> I meant threshold for when we use start using single mode. So it's always either all items or just the hovered one, e.g. if the threshold was "5":

If we do go down this route I would strongly advice against going back to only displaying 1 value if the number of series is over a threshold.  I would suggest displying the _n_ series nearest the cursor if that can be made performant.

**Comment by ivanahuckova** [comment]
Closing this as we are not certain that `multi` tooltip in current form wouldn't be obstacle for users who use Explore to interact with many series. We created https://github.com/grafana/grafana/issues/61616 issue that is specific for this usecase and https://github.com/grafana/grafana/issues/61615 for generic tooltip issue - that `multi` tooltip doesn't work the best with many series.

**Pull request closed without merging**
